File: clients_java/client/src/org/openamq/client/handler/ConnectionCloseMethodHandler.java
package org.openamq.client.handler;

import org.apache.log4j.Logger;
import org.openamq.AMQException;
import org.openamq.AMQConnectionClosedException;
import org.openamq.client.protocol.AMQMethodEvent;
import org.openamq.client.state.AMQState;
import org.openamq.client.state.AMQStateManager;
import org.openamq.client.state.StateAwareMethodListener;
import org.openamq.framing.ConnectionCloseBody;
import org.openamq.framing.ConnectionCloseOkBody;

/**
 * @author Robert Greig (robert.j.greig@jpmorgan.com)
 */
public class ConnectionCloseMethodHandler implements StateAwareMethodListener
{
    private static final Logger _logger = Logger.getLogger(ConnectionCloseMethodHandler.class);

    private static ConnectionCloseMethodHandler _handler = new ConnectionCloseMethodHandler();

    public static ConnectionCloseMethodHandler getInstance()
    {
        return _handler;
    }

    private ConnectionCloseMethodHandler()
    {
    }

    public void methodReceived(AMQStateManager stateManager, AMQMethodEvent evt) throws AMQException
    {
        _logger.debug("ConnectionClose frame received");
        ConnectionCloseBody method = (ConnectionCloseBody) evt.getMethod();

        int errorCode = method.replyCode;
        String reason = method.replyText;

        // TODO: check whether channel id of zero is appropriate
        evt.getProtocolSession().writeFrame(ConnectionCloseOkBody.createAMQFrame((short)0));
        stateManager.changeState(AMQState.CONNECTION_CLOSED);
        if (errorCode != 200)
        {
            _logger.debug("Connection close received with error code " + errorCode);
            throw new AMQConnectionClosedException(errorCode, "Error: " + reason);
        }

        // this actually closes the connection in the case where it is not an error.
        evt.getProtocolSession().closeProtocolSession();
    }
}

File: dotnet/xmsclient/Client/Handler/ConnectionCloseMethodHandler.cs
using System;
using jpmorgan.mina.common;
using log4net;
using OpenAMQ.Framing;
using OpenAMQ.XMS.Client.Protocol;
using OpenAMQ.XMS.Client.State;

namespace OpenAMQ.XMS.Client.Handler
{
    public class ConnectionCloseMethodHandler : IStateAwareMethodListener
    {
        private static readonly ILog _logger = LogManager.GetLogger(typeof(ConnectionCloseMethodHandler));

        private static ConnectionCloseMethodHandler _handler = new ConnectionCloseMethodHandler();

        public static ConnectionCloseMethodHandler getInstance()
        {
            return _handler;
        }

        private ConnectionCloseMethodHandler()
        {
        }

        public void MethodReceived(AMQStateManager stateManager, AMQMethodEvent evt)
        {
            _logger.Debug("ConnectionClose frame received");
            ConnectionCloseBody method = (ConnectionCloseBody) evt.Method;

            int errorCode = method.ReplyCode;
            String reason = method.ReplyText;

            // TODO: check whether channel id of zero is appropriate
            evt.ProtocolSession.WriteFrame(ConnectionCloseOkBody.CreateAMQFrame(0));
            stateManager.ChangeState(AMQState.CONNECTION_CLOSED);
            if (errorCode != 200)
            {
                _logger.Debug("Connection close received with error code " + errorCode);
                throw new AMQConnectionClosedException(errorCode, "Error: " + reason);
            }

            // this actually closes the connection in the case where it is not an error.
            evt.ProtocolSession.CloseProtocolSession();
        }
    }
}
